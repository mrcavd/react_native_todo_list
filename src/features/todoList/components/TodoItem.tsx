import React from "react";
import { Text, TouchableOpacity, View } from "react-native";
import { TODO_STYLE } from "../../../constants";
import { TodoItemType } from "../../../types";
import { ActionButton } from "../../universal";

interface TodoItemProps {
    item: TodoItemType;
    editCallback: (id: string) => void;
    removeCallback: (id: string) => void;
}

const TodoItem: React.FC<TodoItemProps> = ({
    item,
    editCallback,
    removeCallback,
}) => {
    const editOnPress = () => {
        editCallback(item.id);
    };

    const removeOnPress = () => {
        removeCallback(item.id);
    };

    return (
        <TouchableOpacity
            key={item.id}
            onPress={editOnPress}
            style={TODO_STYLE.todoItemContainer}
            testID="todo_item"
        >
            <View style={TODO_STYLE.todoItemInfoContainer}>
                <View style={TODO_STYLE.todoItemStatus} />
                <View style={TODO_STYLE.todoItemTextContainer}>
                    <Text style={TODO_STYLE.todoItemText} testID="label_text">
                        {item.label}
                    </Text>
                </View>
            </View>
            <ActionButton
                label={"REMOVE"}
                onPress={removeOnPress}
                buttonStyle={TODO_STYLE.removeButton}
                textStyle={TODO_STYLE.todoItemText}
            />
        </TouchableOpacity>
    );
};

export default TodoItem;
