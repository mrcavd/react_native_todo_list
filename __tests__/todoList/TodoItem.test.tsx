import React from "react";
import renderer from "react-test-renderer";
import { jest, describe, it, expect } from "@jest/globals";
import TodoItem from "../../src/features/todoList/components/TodoItem";

describe("LandmarkOfferTopHat", () => {
    it("TodoItem renders correctly", () => {
        const item = { id: "1", label: "Test Todo" };
        const tree = renderer
            .create(
                <TodoItem
                    item={item}
                    editCallback={() => {}}
                    removeCallback={() => {}}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("Edit callback is called when TodoItem is pressed", () => {
        const item = { id: "unique_id_1", label: "Test Todo" };
        const editCallbackMock = jest.fn();
        const component = renderer.create(
            <TodoItem
                item={item}
                editCallback={editCallbackMock}
                removeCallback={() => {}}
            />
        );
        const todoItem = component.root.findByProps({ testID: "todo_item" });

        todoItem.props.onPress();
        expect(editCallbackMock).toHaveBeenCalledWith("unique_id_1");
    });

    it('Remove callback is called when "REMOVE" button is pressed', () => {
        const item = { id: "unique_id_1", label: "Test Todo" };
        const removeCallbackMock = jest.fn();
        const component = renderer.create(
            <TodoItem
                item={item}
                editCallback={() => {}}
                removeCallback={removeCallbackMock}
            />
        );
        const removeButton = component.root.findByProps({
            testID: "remove_button",
        });

        removeButton.props.onPress();
        expect(removeCallbackMock).toHaveBeenCalledWith("unique_id_1");
    });
});
