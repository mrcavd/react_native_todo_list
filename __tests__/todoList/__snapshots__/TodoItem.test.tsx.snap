// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandmarkOfferTopHat TodoItem renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderRadius": 20,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 15,
      "marginHorizontal": 10,
      "opacity": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 20,
    }
  }
  testID="todo_item"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "75%",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#305ABB",
          "borderRadius": 25,
          "height": 25,
          "width": 25,
        }
      }
    />
    <View
      style={
        {
          "flex": 1,
          "marginHorizontal": 5,
        }
      }
    >
      <Text
        style={
          {
            "color": "#222222",
            "fontSize": 16,
            "marginHorizontal": 10,
          }
        }
        testID="label_text"
      >
        Test Todo
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "paddingHorizontal": 5,
        "paddingVertical": 5,
      }
    }
    testID="remove_button"
  >
    <Text
      style={
        {
          "color": "#222222",
          "fontSize": 16,
          "marginHorizontal": 10,
        }
      }
    >
      REMOVE
    </Text>
  </View>
</View>
`;
